File: scripts/scrape-indiamart-images.py
#!/usr/bin/env python3
"""Scrape IndiaMART product pack shots using a logged-in session cookie file.

1) Log in at buyer.indiamart.com / dir.indiamart.com in a browser
2) Export cookies to /tmp/indiamart-cookies.json (Playwright storage_state)
   OR pass --cookie-header "name=value; ..."
3) Run:
   .venv-scrape/bin/python scripts/scrape-indiamart-images.py --cookie-header-file /tmp/im-cookies.txt

Cookies are never written into the repo.
"""

from __future__ import annotations

import argparse
import csv
import json
import re
import time
import urllib.parse
import urllib.request
from pathlib import Path

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parents[1]
CSV_PATH = ROOT / "docs" / "seller-product-catalog.csv"
OUT_DIR = ROOT / "public" / "media" / "catalog"
UA = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
)


def slugify(text: str) -> str:
    text = text.lower().replace("&", " and ")
    return re.sub(r"[^a-z0-9]+", "-", text).strip("-")[:80]


def image_key(row: dict) -> str:
    title = row.get("title") or ""
    pack = (row.get("pack_size") or "").strip()
    if pack:
        title = re.sub(re.escape(pack), "", title, flags=re.I)
    title = re.sub(
        r"\b(\d+(\.\d+)?\s?(kg|g|ml|l|pcs?|pack|bags?|dozen|piece|tray|set|box|tin|jar|cup|bottle|can)\b)",
        "",
        title,
        flags=re.I,
    )
    title = re.sub(r"\s+", " ", title).strip(" -")
    brand = (row.get("brand") or "").strip()
    if brand and title.lower().startswith(brand.lower()):
        base = title
    else:
        base = f"{brand} {title}".strip() or row["title"]
    return slugify(base)


def search_query(row: dict) -> str:
    brand = (row.get("brand") or "").strip()
    title = row.get("title") or ""
    pack = (row.get("pack_size") or "").strip()
    if pack:
        title = title.replace(pack, "").strip()
    skip = {"local", "local fresh", "local premium", "imported", "assorted"}
    words = " ".join(title.split()[:6])
    if brand and brand.lower() not in skip and brand.lower() not in words.lower():
        return f"{brand} {words}".strip()
    return words or row["title"]


def candidate_slugs(row: dict) -> list[str]:
    q = search_query(row)
    words = [w for w in q.split() if len(w) > 2]
    out: list[str] = []
    for s in (
        slugify(q),
        slugify(" ".join(words[:2])) if len(words) >= 2 else "",
        slugify(words[-1]) if words else "",
    ):
        if s and s not in out:
            out.append(s)
    return out


def pick_image(html: str, query: str) -> str | None:
    imgs = re.findall(
        r"https?://[0-9]?\.?imimg\.com[^\"'\s>]+\.(?:jpg|jpeg|png|webp)", html, re.I
    )
    tokens = [t for t in query.lower().split() if len(t) > 2]
    best, best_score = None, -1
    for u in dict.fromkeys(imgs):
        url = u.replace("http://", "https://")
        if re.search(
            r"logo|icon|sprite|placeholder|250x250|125x125|whatsapp-image", url, re.I
        ):
            continue
        score = 0
        if re.search(r"1000x1000", url, re.I):
            score += 300
        if re.search(r"500x500", url, re.I):
            score += 120
        low = url.lower()
        for t in tokens:
            if t in low:
                score += 90
        if score > best_score:
            best_score, best = score, url
    if best_score >= 90:
        return best
    for u in dict.fromkeys(imgs):
        url = u.replace("http://", "https://")
        if not re.search(r"logo|icon|sprite|250x250|125x125", url, re.I):
            return url
    return None


def parse_cookie_header(header: str) -> list[dict]:
    cookies = []
    for part in header.split(";"):
        part = part.strip()
        if not part or "=" not in part:
            continue
        name, value = part.split("=", 1)
        cookies.append(
            {
                "name": name.strip(),
                "value": value.strip(),
                "domain": ".indiamart.com",
                "path": "/",
            }
        )
    return cookies


def download(url: str, key: str) -> str:
    req = urllib.request.Request(
        url,
        headers={"User-Agent": UA, "Referer": "https://www.indiamart.com/", "Accept": "image/*"},
    )
    with urllib.request.urlopen(req, timeout=40) as resp:
        data = resp.read()
        ctype = (resp.headers.get_content_type() or "").lower()
    if len(data) < 2000:
        raise ValueError("image too small")
    ext = ".jpg"
    if "png" in ctype or url.lower().endswith(".png"):
        ext = ".png"
    elif "webp" in ctype or ".webp" in url.lower():
        ext = ".webp"
    dest = OUT_DIR / f"{key}{ext}"
    for old in OUT_DIR.glob(f"{key}.*"):
        if old != dest:
            old.unlink(missing_ok=True)
    dest.write_bytes(data)
    return f"/media/catalog/{dest.name}"


def write_csv(rows: list[dict]) -> None:
    fields = [
        "category",
        "title",
        "brand",
        "pack_size",
        "unit",
        "price_inr",
        "inventory",
        "short_description",
        "aliases",
        "image_path",
        "image_url",
        "image_source",
    ]
    with CSV_PATH.open("w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=fields, extrasaction="ignore")
        w.writeheader()
        for r in rows:
            w.writerow({k: r.get(k, "") for k in fields})


def find_image(page, row: dict) -> tuple[str, str] | None:
    """Use in-page fetch (credentials/cookies) — same method as browser DOM scrape."""
    q = search_query(row)
    urls = [f"https://dir.indiamart.com/impcat/{s}.html" for s in candidate_slugs(row)]
    urls.append(f"https://dir.indiamart.com/search.mp?ss={urllib.parse.quote(q)}")

    result = page.evaluate(
        """async ({ urls, query }) => {
      const pick = (imgs, query) => {
        const tokens = query.toLowerCase().split(/\\s+/).filter(t => t.length > 2);
        let best = null, bestScore = -1;
        for (const u of imgs) {
          const url = String(u).replace(/^http:/,'https:');
          if (/logo|icon|sprite|placeholder|250x250|125x125|whatsapp-image/i.test(url)) continue;
          let score = 0;
          if (/1000x1000/i.test(url)) score += 300;
          if (/500x500/i.test(url)) score += 120;
          const low = url.toLowerCase();
          for (const t of tokens) if (low.includes(t)) score += 90;
          if (score > bestScore) { bestScore = score; best = url; }
        }
        if (bestScore >= 90) return best;
        for (const u of imgs) {
          const url = String(u).replace(/^http:/,'https:');
          if (!/logo|icon|sprite|250x250|125x125/i.test(url)) return url;
        }
        return null;
      };
      for (const url of urls) {
        try {
          const res = await fetch(url, { credentials: 'include', headers: { 'Accept': 'text/html' } });
          if (!res.ok) continue;
          const html = await res.text();
          if (/page not found/i.test(html) && html.length < 20000) continue;
          const imgs = [...html.matchAll(/https?:\\/\\/[0-9]?\\.?imimg\\.com[^"'\\s>]+\\.(?:jpg|jpeg|png|webp)/gi)].map(m => m[0]);
          const img = pick([...new Set(imgs)], query);
          if (img) return { image: img, via: url.includes('/impcat/') ? 'impcat' : 'search' };
        } catch (e) {}
      }
      return null;
    }""",
        {"urls": urls, "query": q},
    )
    if result and result.get("image"):
        return result["image"], result.get("via") or "search"
    return None


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--cookie-header-file", required=True, help="File with document.cookie text")
    ap.add_argument("--limit", type=int, default=0)
    ap.add_argument("--force", action="store_true")
    args = ap.parse_args()

    cookie_header = Path(args.cookie_header_file).read_text(encoding="utf-8").strip()
    if not cookie_header:
        raise SystemExit("empty cookie header file")

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    rows = list(csv.DictReader(CSV_PATH.open(encoding="utf-8")))
    by_key: dict[str, dict] = {}
    for r in rows:
        by_key.setdefault(image_key(r), r)

    keys = list(by_key)
    if not args.force:
        keys = [
            k
            for k in keys
            if not (
                by_key[k].get("image_source") == "indiamart"
                and by_key[k].get("image_path")
                and (ROOT / "public" / by_key[k]["image_path"].lstrip("/")).exists()
            )
        ]
    if args.limit:
        keys = keys[: args.limit]

    print(f"Unique products to scrape: {len(keys)} (of {len(by_key)})")
    cache: dict[str, dict] = {}
    ok = fail = 0

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(user_agent=UA, locale="en-IN", viewport={"width": 1400, "height": 900})
        context.add_cookies(parse_cookie_header(cookie_header))
        page = context.new_page()
        page.goto("https://dir.indiamart.com/", wait_until="domcontentloaded", timeout=60000)
        page.wait_for_timeout(1000)
        # verify login-ish cookie present
        has_iss = page.evaluate("document.cookie.includes('im_iss=') || document.cookie.includes('ImeshVisitor=')")
        print(f"session_cookie_present={has_iss}")

        for i, key in enumerate(keys, 1):
            row = by_key[key]
            print(f"[{i}/{len(keys)}] {row['title'][:55]} …", end=" ", flush=True)
            try:
                found = find_image(page, row)
                if not found:
                    raise RuntimeError("no image")
                url, via = found
                path = download(url, key)
                cache[key] = {
                    "image_path": path,
                    "image_url": url,
                    "image_source": "indiamart",
                }
                ok += 1
                print(f"✓ {via}")
            except Exception as exc:
                fail += 1
                print(f"✗ {exc}")
            if i % 15 == 0:
                for r in rows:
                    m = cache.get(image_key(r))
                    if m:
                        r.update(m)
                write_csv(rows)
            time.sleep(0.05)

        browser.close()

    for r in rows:
        m = cache.get(image_key(r))
        if m:
            r.update(m)
    write_csv(rows)
    filled = sum(1 for r in rows if r.get("image_source") == "indiamart")
    print(f"\nDone. scraped_ok={ok} failed={fail} csv_indiamart_rows={filled}/{len(rows)}")


if __name__ == "__main__":
    main()

File: scripts/scrape-indiamart-images.mjs
#!/usr/bin/env node
/**
 * Replace garbage stock photos with IndiaMART product pack shots.
 *
 * Strategy:
 * 1) Open dir.indiamart.com/impcat/{slug}.html when possible
 * 2) Else open search.mp?ss=... and dismiss login modal
 * 3) Pull largest 5.imimg.com product image from DOM
 * 4) Download to public/media/catalog and update CSV
 *
 * Usage: node scripts/scrape-indiamart-images.mjs [--limit N] [--force]
 */
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { chromium } from 'playwright'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const ROOT = path.resolve(__dirname, '..')
const CSV_PATH = path.join(ROOT, 'docs/seller-product-catalog.csv')
const OUT_DIR = path.join(ROOT, 'public/media/catalog')

const args = new Set(process.argv.slice(2))
const force = args.has('--force')
const limitIdx = process.argv.indexOf('--limit')
const limit = limitIdx >= 0 ? Number(process.argv[limitIdx + 1]) : Infinity

function parseCsv(text) {
  const rows = []
  let i = 0
  let field = ''
  let row = []
  let inQuotes = false
  while (i < text.length) {
    const c = text[i]
    if (inQuotes) {
      if (c === '"' && text[i + 1] === '"') {
        field += '"'
        i += 2
        continue
      }
      if (c === '"') {
        inQuotes = false
        i++
        continue
      }
      field += c
      i++
      continue
    }
    if (c === '"') {
      inQuotes = true
      i++
      continue
    }
    if (c === ',') {
      row.push(field)
      field = ''
      i++
      continue
    }
    if (c === '\n' || c === '\r') {
      if (c === '\r' && text[i + 1] === '\n') i++
      row.push(field)
      if (row.some((x) => x !== '')) rows.push(row)
      row = []
      field = ''
      i++
      continue
    }
    field += c
    i++
  }
  if (field.length || row.length) {
    row.push(field)
    rows.push(row)
  }
  const header = rows[0]
  return rows.slice(1).map((r) => {
    const obj = {}
    header.forEach((h, idx) => {
      obj[h] = r[idx] ?? ''
    })
    return obj
  })
}

function toCsv(rows) {
  const header = [
    'category',
    'title',
    'brand',
    'pack_size',
    'unit',
    'price_inr',
    'inventory',
    'short_description',
    'aliases',
    'image_path',
    'image_url',
    'image_source',
  ]
  const esc = (v) => {
    const s = String(v ?? '')
    if (/[",\n\r]/.test(s)) return `"${s.replace(/"/g, '""')}"`
    return s
  }
  return [header.join(','), ...rows.map((r) => header.map((h) => esc(r[h])).join(','))].join('\n') + '\n'
}

function slugify(text) {
  return String(text || '')
    .toLowerCase()
    .replace(/&/g, ' and ')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
    .slice(0, 80)
}

function imageKey(row) {
  let title = row.title || ''
  const pack = (row.pack_size || '').trim()
  if (pack) title = title.replace(new RegExp(pack.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'ig'), '')
  title = title
    .replace(/\b(\d+(\.\d+)?\s?(kg|g|ml|l|pcs?|pack|bags?|dozen|piece|tray|set|box|tin|jar|cup|bottle|can))\b/gi, '')
    .replace(/\s+/g, ' ')
    .trim()
  const brand = (row.brand || '').trim()
  const base =
    brand && title.toLowerCase().startsWith(brand.toLowerCase())
      ? title
      : `${brand} ${title}`.trim() || row.title
  return slugify(base)
}

function searchQuery(row) {
  const brand = (row.brand || '').trim()
  let title = row.title || ''
  const pack = (row.pack_size || '').trim()
  if (pack) title = title.replace(pack, '').trim()
  const skip = new Set(['local', 'local fresh', 'local premium', 'imported', 'assorted'])
  const words = title.split(/\s+/).filter(Boolean).slice(0, 6).join(' ')
  if (brand && !skip.has(brand.toLowerCase()) && !words.toLowerCase().includes(brand.toLowerCase())) {
    return `${brand} ${words}`.trim()
  }
  return words || row.title
}

function candidateSlugs(row) {
  const q = searchQuery(row)
  const brand = (row.brand || '').trim()
  const out = new Set()
  out.add(slugify(q))
  // brand + first meaningful product words
  const words = q.split(/\s+/).filter((w) => w.length > 2)
  if (words.length >= 2) out.add(slugify(words.slice(0, 3).join(' ')))
  if (words.length >= 2) out.add(slugify(words.slice(0, 2).join(' ')))
  if (brand && !['local', 'local fresh', 'local premium', 'imported', 'assorted'].includes(brand.toLowerCase())) {
    out.add(slugify(brand))
  }
  return [...out].filter(Boolean)
}

function scoreImage(img, query) {
  const src = img.src || ''
  if (!/imimg\.com/i.test(src)) return -1
  if (/logo|icon|sprite|placeholder|noimage|default-img/i.test(src)) return -1
  const alt = (img.alt || '').toLowerCase()
  const q = query.toLowerCase()
  const tokens = q.split(/\s+/).filter((t) => t.length > 2)
  let score = (img.w || 0) + (img.h || 0)
  for (const t of tokens) {
    if (alt.includes(t)) score += 200
    if (src.toLowerCase().includes(t)) score += 80
  }
  if (/1000x1000|500x500|800x800/i.test(src)) score += 100
  if (/250x250|125x125/i.test(src)) score -= 150
  return score
}

async function dismissModals(page) {
  for (const sel of [
    'button:has-text("Close")',
    '[aria-label="Close"]',
    'button:has-text("×")',
    '.close',
  ]) {
    try {
      const el = page.locator(sel).first()
      if (await el.isVisible({ timeout: 500 })) await el.click({ timeout: 1000 }).catch(() => {})
    } catch {
      /* ignore */
    }
  }
  await page.keyboard.press('Escape').catch(() => {})
}

async function collectImages(page) {
  return page.evaluate(() => {
    return [...document.querySelectorAll('img')]
      .map((i) => ({
        src: i.currentSrc || i.src || i.getAttribute('data-src') || '',
        alt: i.alt || '',
        w: i.naturalWidth || i.width || 0,
        h: i.naturalHeight || i.height || 0,
      }))
      .filter((x) => x.src && /^https?:/i.test(x.src))
  })
}

async function findImageForProduct(page, row) {
  const query = searchQuery(row)
  const slugs = candidateSlugs(row)

  for (const slug of slugs) {
    const url = `https://dir.indiamart.com/impcat/${slug}.html`
    try {
      const resp = await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 45000 })
      await dismissModals(page)
      await page.waitForTimeout(1200)
      const title = await page.title()
      if (resp && resp.status() >= 400) continue
      if (/not found|404/i.test(title)) continue
      const imgs = await collectImages(page)
      const ranked = imgs
        .map((img) => ({ img, score: scoreImage(img, query) }))
        .filter((x) => x.score > 100)
        .sort((a, b) => b.score - a.score)
      if (ranked[0]) {
        return { url: ranked[0].img.src, via: `impcat:${slug}`, alt: ranked[0].img.alt }
      }
    } catch {
      /* try next */
    }
  }

  // Search fallback
  const searchUrl = `https://dir.indiamart.com/search.mp?ss=${encodeURIComponent(query)}`
  try {
    await page.goto(searchUrl, { waitUntil: 'domcontentloaded', timeout: 45000 })
    await dismissModals(page)
    await page.waitForTimeout(1500)
    await dismissModals(page)
    // Prefer clicking first proddetail-looking link if present
    const detail = page.locator('a[href*="/proddetail/"]').first()
    if (await detail.count()) {
      await detail.click({ timeout: 5000 }).catch(() => {})
      await page.waitForTimeout(1500)
      await dismissModals(page)
    }
    const imgs = await collectImages(page)
    const ranked = imgs
      .map((img) => ({ img, score: scoreImage(img, query) }))
      .filter((x) => x.score > 80)
      .sort((a, b) => b.score - a.score)
    if (ranked[0]) {
      return { url: ranked[0].img.src, via: 'search', alt: ranked[0].img.alt }
    }
  } catch {
    /* ignore */
  }

  // Direct Google-less proddetail guess via IndiaMART search page cards sometimes lazy-load
  return null
}

async function downloadImage(url, destBase) {
  const res = await fetch(url, {
    headers: {
      'User-Agent':
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      Referer: 'https://www.indiamart.com/',
      Accept: 'image/avif,image/webp,image/apng,image/*,*/*;q=0.8',
    },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  const buf = Buffer.from(await res.arrayBuffer())
  if (buf.length < 2000) throw new Error('image too small')
  const ctype = (res.headers.get('content-type') || '').toLowerCase()
  let ext = '.jpg'
  if (ctype.includes('png')) ext = '.png'
  else if (ctype.includes('webp')) ext = '.webp'
  else if (ctype.includes('jpeg') || ctype.includes('jpg')) ext = '.jpg'
  else {
    const m = url.toLowerCase().match(/\.(jpg|jpeg|png|webp)(?:\?|$)/)
    if (m) ext = m[1] === 'jpeg' ? '.jpg' : `.${m[1]}`
  }
  const dest = `${destBase}${ext}`
  // remove prior extensions for this key
  for (const oldExt of ['.jpg', '.jpeg', '.png', '.webp', '.avif', '.gif']) {
    const p = `${destBase}${oldExt}`
    if (p !== dest && fs.existsSync(p)) fs.unlinkSync(p)
  }
  fs.writeFileSync(dest, buf)
  return { dest, ext, bytes: buf.length }
}

async function main() {
  fs.mkdirSync(OUT_DIR, { recursive: true })
  const rows = parseCsv(fs.readFileSync(CSV_PATH, 'utf8'))
  const byKey = new Map()
  for (const row of rows) {
    const key = imageKey(row)
    if (!byKey.has(key)) byKey.set(key, row)
  }

  let keys = [...byKey.keys()]
  if (!force) {
    keys = keys.filter((key) => {
      const row = byKey.get(key)
      const existing = row.image_source === 'indiamart' && row.image_path
      const fileOk =
        existing && fs.existsSync(path.join(ROOT, 'public', row.image_path.replace(/^\//, '')))
      return !fileOk
    })
  }
  keys = keys.slice(0, Number.isFinite(limit) ? limit : keys.length)

  console.log(`Unique products to scrape: ${keys.length} (of ${byKey.size})`)

  const browser = await chromium.launch({ headless: true })
  const context = await browser.newContext({
    userAgent:
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    viewport: { width: 1400, height: 900 },
    locale: 'en-IN',
  })
  const page = await context.newPage()

  const cache = new Map() // key -> meta
  let ok = 0
  let fail = 0

  for (let i = 0; i < keys.length; i++) {
    const key = keys[i]
    const row = byKey.get(key)
    process.stdout.write(`[${i + 1}/${keys.length}] ${row.title.slice(0, 55)} … `)
    try {
      const found = await findImageForProduct(page, row)
      if (!found?.url) throw new Error('no image in DOM')
      const destBase = path.join(OUT_DIR, key)
      const { dest, bytes } = await downloadImage(found.url, destBase)
      const rel = `/media/catalog/${path.basename(dest)}`
      cache.set(key, {
        image_path: rel,
        image_url: found.url,
        image_source: 'indiamart',
      })
      ok++
      console.log(`✓ ${path.basename(dest)} (${bytes}b via ${found.via})`)
    } catch (err) {
      fail++
      console.log(`✗ ${err.message || err}`)
    }
    // checkpoint CSV every 10
    if ((i + 1) % 10 === 0) {
      applyCache(rows, cache)
      fs.writeFileSync(CSV_PATH, toCsv(rows))
    }
  }

  applyCache(rows, cache)
  fs.writeFileSync(CSV_PATH, toCsv(rows))
  await browser.close()

  const filled = rows.filter((r) => r.image_source === 'indiamart').length
  console.log(`\nDone. scraped_ok=${ok} failed=${fail} csv_indiamart_rows=${filled}/${rows.length}`)
}

function applyCache(rows, cache) {
  for (const row of rows) {
    const key = imageKey(row)
    const meta = cache.get(key)
    if (!meta) continue
    row.image_path = meta.image_path
    row.image_url = meta.image_url
    row.image_source = meta.image_source
  }
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
